keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,601 @@
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a feature to protect Secrets/ConfigMaps while it is in use. Currently, users can delete a secret that is being used by other resources, like Pods, PVs, and VolumeSnapshots. This may have negative impact on the resouces using the secret and it may result in data loss.
+
+For example, one of the worst case scenarios is deletion of a Node Stage Secret for CSI volumes while it is still in-use.
+If it happens, a CSI driver for the volume can't unstage the volume due to the lack of the Secret,
+as a result, the volume can't be detached from the node and the volume can't even be deleted.
+This issue will easily happen if a PVC for the volume and a Secret for the volume exist in the same namespace, and a user requests to delete the namespace, which starts deletion of all the resources in the namespace.
+When the Secret is deleted before the PVC is deleted, this issue happens.
+
+Also, ConfigMaps can be deleted while it is in use, which may lead to an unexpected behavior in applications.
+
+Similar features for protecting PV and PVC already exist as [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md).
+
+
+## Motivation
+
+This feature aims to protect Secrets/ConfigMaps from being deleted while they are in-use.
+Secrets can be used by below ways:
+- From Pod:
+  - [Mounted as data volumes](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod)
+  - [Exposed as environment variables](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables)
```

Comment:
shouldn't the kubelet safely coast on existing data if in-use secrets/configmaps are removed?

Reply:
Interesting point. It will depend on what we define the "safe" behavior.

As I tested below, the value of the secret in the pod is updated properly and exists even after the secret is deleted. Therefore, when seen from an application in a pod, it may already be "safe".

On the other hand, when seen from application developers, it may not be safe that operators can erroneously delete a secret while a pod is still using it. It will hide the invalid status that pod will fail on the next pod restart, until it actually is restarted.

In this sense, current spec of protecting just from PodSpec in Pod might not be enough, and we may need to consider also adding PodTemplateSpec in Deployment. Because there may be a short periods of time between pod is deleted and restarted that operator can delete a secret that should be regarded as in-use. Any thoughts?

[Current behavior of secret in pod]

1. Create secret and mount it on the pod (In console A)
```bash
kubectl create secret generic test-secret --from-literal='username=my-app' --from-literal='password=39528$vdg7Jb'
kubectl apply -f https://k8s.io/examples/pods/inject/secret-pod.yaml
```

2. Check the value of the secret from the pod (In console B)
```bash
kubectl exec -i -t secret-test-pod -- /bin/bash
/# echo "$( cat /etc/secret-volume/username )"
my-app
```

3. Update the value of the secret (In console A)
```bash
kubectl create secret generic test-secret --from-literal='username=my-app2' --from-literal='password=39528$vdg7Jb' --dry-run=client -o yaml | kubectl apply -f -
```

4. Check the value of the secret from the pod (In console B)
```bash
/# echo "$( cat /etc/secret-volume/username )"
my-app2
```

5. Delete the secret (In console A)
```bash
kubectl delete secret test-secret
secret "test-secret" deleted
```

6. Check the value of the secret from the pod (In console B)
```bash
echo "$( cat /etc/secret-volume/username )"
my-app2
```

Reply:
Actually, I don't have a strong objection in changing used-from-pod secret to out-of-scope, because my main concern is protecting used-from-PVC/snapshot secret. This issue is really a serious problem that needs to be handled properly.

Reply:
> my main concern is protecting used-from-PVC/snapshot secret. This issue is really a serious problem that needs to be handled properly.

I still think placing the secrets required to administrate a volume for a PVC in the same namespace as the PVC is not a good approach. Would placing those secrets in a dedicated namespace separate from the PVC address the main issue this KEP is trying to resolve?

> Updated to include Deployment, StatefulSet, Daemonset, and CronJob to the scope, which use podTemplateSpec (Intentionally excluded Job, because the existence of Job doesn't necessary mean that a Pod will be created from the template in the future).

I don't think we should spider out to all the pod-template-containing objects. I'm pretty skeptical about the reliability/performance of this feature already, relative to the benefit.